Source/Core/DolphinQt2/NetPlay/NetPlayDialog.cpp

```diff
@@ -0,0 +1,549 @@
+  settings.m_EXIDevice[0] = instance.m_EXIDevice[0];
+  settings.m_EXIDevice[1] = instance.m_EXIDevice[1];
+
+  auto* server = QtNetPlay::GetServer();
+  server->SetNetSettings(settings);
+  server->StartGame();
+}
+
+void NetPlayDialog::OnMD5Combo(int index)
+{
+  std::string identifier;
+
+  switch (index)
+  {
+  case 0:
+    return;
+  case 1:  // Current game
+    identifier = GameFile(QString::fromStdString(m_current_game)).GetUniqueID().toStdString();
+    break;
+  case 2:  // Other game
+  {
+    GameListDialog gld(this);
+
+    if (gld.exec() == QDialog::Accepted)
+    {
+      identifier = GameFile(gld.GetGamePath()).GetUniqueID().toStdString();
+      break;
+    }
+    else
+    {
+      m_md5_box->setCurrentIndex(0);
+      return;
+    }
+  }
+  case 3:  // SD Card
+    identifier = WII_SDCARD;
+    break;
+  }
+
+  m_md5_box->setEnabled(false);
+  m_md5_box->setCurrentIndex(0);
+  QtNetPlay::GetServer()->ComputeMD5(identifier);
+}
+
+void NetPlayDialog::reject()
+{
+  if (QMessageBox::question(this, tr("Confirmation"),
+                            tr("Are you sure you want to quit NetPlay?")) == QMessageBox::Yes)
+  {
+    QtNetPlay::Quit();
+    m_chat_edit->clear();
+    m_chat_type_edit->clear();
+    QDialog::reject();
+  }
+}
+
+void NetPlayDialog::show()
+{
+  m_room_box->clear();
+
+  bool is_hosting = QtNetPlay::GetServer() != nullptr;
+
+  if (is_hosting)
+  {
+    if (QtNetPlay::UseTraversal())
+      m_room_box->addItem(tr("Room ID"));
+
+    for (const auto& iface : QtNetPlay::GetServer()->GetInterfaceSet())
+      m_room_box->addItem(QString::fromStdString(iface));
+  }
+
+  m_start_button->setHidden(!is_hosting);
+  m_save_sd_box->setHidden(!is_hosting);
+  m_load_wii_box->setHidden(!is_hosting);
+  m_buffer_size_box->setHidden(!is_hosting);
+  m_buffer_label->setHidden(!is_hosting);
+  m_kick_button->setHidden(!is_hosting);
+  m_assign_ports_button->setHidden(!is_hosting);
+  m_md5_box->setHidden(!is_hosting);
+  m_room_box->setHidden(!is_hosting);
+  m_hostcode_label->setHidden(!is_hosting);
+  m_hostcode_copy_button->setHidden(!is_hosting);
+  m_game_button->setEnabled(is_hosting);
+  m_kick_button->setEnabled(false);
+
+  QDialog::show();
+}
+
+void NetPlayDialog::UpdateGUI()
+{
+  if (!isVisible())
+    return;
+
+  // Update player list
+  static std::vector<int> player_ids;
+  std::string tmp;
+
+  QtNetPlay::GetClient()->GetPlayerList(tmp, player_ids);
+
+  std::istringstream ss(tmp);
+  int row = m_players_list->currentRow();
+  m_players_list->clear();
+
+  while (std::getline(ss, tmp))
+  {
+    auto text = QString::fromStdString(tmp);
+    if (!text.isEmpty())
+    {
+      QListWidgetItem* item = new QListWidgetItem(text);
+      if (player_ids.size() > 0 && !text.startsWith(QStringLiteral("Ping:")) &&
+          !text.startsWith(QStringLiteral("Status:")))
+      {
+        item->setData(Qt::UserRole, player_ids[0]);
+        player_ids.erase(player_ids.begin());
```

Comment:
Erasing from the beginning of a vector is expensive. Can't you increment an index into player_ids?